File: parser.cpp
// system includes
#include <map>
#include <memory>

// includes
#include "parser.h"
#include "basic_parser/parser.h"
#include "basic_parser/malformed_request.h"
#include "validator.h"
#include "request_type_parser.h"

namespace generic_protocol
{

namespace parser
{

using ::basic_parser::parser::get_value_or_throw;
using ::basic_parser::parser::get_value_or_throw_t;

// enums

void get_value_or_throw( ErrorResponse_type_e * res, const std::string & key, const generic_request::Request & r )
{
    uint32_t res_i;

    get_value_or_throw( & res_i, key, r );

    * res = static_cast<ErrorResponse_type_e>( res_i );
}

// objects

void get_value_or_throw( SessionInfo * res, const std::string & prefix, const generic_request::Request & r )
{
    get_value_or_throw( & res->user_id, prefix + ".USER_ID", r );
    get_value_or_throw( & res->start_time, prefix + ".START_TIME", r );
    get_value_or_throw( & res->expiration_time, prefix + ".EXPIRATION_TIME", r );
}

// base messages

void get_value_or_throw( ForwardMessage * res, const generic_request::Request & r )
{
}

void get_value_or_throw( BackwardMessage * res, const generic_request::Request & r )
{
}

void get_value_or_throw( Request * res, const generic_request::Request & r )
{
    get_value_or_throw( & res->session_id, "SESSION_ID", r );
}

// messages

void get_value_or_throw( ErrorResponse * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<BackwardMessage*>( res ), r );

    get_value_or_throw( & res->type, "TYPE", r );
    get_value_or_throw( & res->descr, "DESCR", r );
}

void get_value_or_throw( AuthenticateRequest * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<ForwardMessage*>( res ), r );

    get_value_or_throw( & res->user_login, "USER_LOGIN", r );
    get_value_or_throw( & res->password, "PASSWORD", r );
}

void get_value_or_throw( AuthenticateAltRequest * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<ForwardMessage*>( res ), r );

    get_value_or_throw( & res->user_id, "USER_ID", r );
    get_value_or_throw( & res->password, "PASSWORD", r );
}

void get_value_or_throw( AuthenticateResponse * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<BackwardMessage*>( res ), r );

    get_value_or_throw( & res->session_id, "SESSION_ID", r );
}

void get_value_or_throw( CloseSessionRequest * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<ForwardMessage*>( res ), r );

    get_value_or_throw( & res->session_id, "SESSION_ID", r );
}

void get_value_or_throw( CloseSessionResponse * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<BackwardMessage*>( res ), r );

}

void get_value_or_throw( GetUserIdRequest * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<Request*>( res ), r );

    get_value_or_throw( & res->user_login, "USER_LOGIN", r );
}

void get_value_or_throw( GetUserIdResponse * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<BackwardMessage*>( res ), r );

    get_value_or_throw( & res->user_id, "USER_ID", r );
}

void get_value_or_throw( GetSessionInfoRequest * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<Request*>( res ), r );

    get_value_or_throw( & res->id, "ID", r );
}

void get_value_or_throw( GetSessionInfoResponse * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<BackwardMessage*>( res ), r );

    get_value_or_throw( & res->session_info, "SESSION_INFO", r );
}

// to object

Object * to_ErrorResponse( const generic_request::Request & r )
{
    std::unique_ptr<ErrorResponse> res( new ErrorResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_AuthenticateRequest( const generic_request::Request & r )
{
    std::unique_ptr<AuthenticateRequest> res( new AuthenticateRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_AuthenticateAltRequest( const generic_request::Request & r )
{
    std::unique_ptr<AuthenticateAltRequest> res( new AuthenticateAltRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_AuthenticateResponse( const generic_request::Request & r )
{
    std::unique_ptr<AuthenticateResponse> res( new AuthenticateResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_CloseSessionRequest( const generic_request::Request & r )
{
    std::unique_ptr<CloseSessionRequest> res( new CloseSessionRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_CloseSessionResponse( const generic_request::Request & r )
{
    std::unique_ptr<CloseSessionResponse> res( new CloseSessionResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetUserIdRequest( const generic_request::Request & r )
{
    std::unique_ptr<GetUserIdRequest> res( new GetUserIdRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetUserIdResponse( const generic_request::Request & r )
{
    std::unique_ptr<GetUserIdResponse> res( new GetUserIdResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetSessionInfoRequest( const generic_request::Request & r )
{
    std::unique_ptr<GetSessionInfoRequest> res( new GetSessionInfoRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetSessionInfoResponse( const generic_request::Request & r )
{
    std::unique_ptr<GetSessionInfoResponse> res( new GetSessionInfoResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

// to forward message

basic_parser::Object* to_forward_message( const generic_request::Request & r )
{
    auto type = detect_request_type( r );

    typedef request_type_e KeyType;

    typedef Object* (*PPMF)( const generic_request::Request & r );

#define HANDLER_MAP_ENTRY(_v)       { KeyType::_v,    & to_##_v }

    static const std::map<KeyType, PPMF> funcs =
    {
        HANDLER_MAP_ENTRY( ErrorResponse ),
        HANDLER_MAP_ENTRY( AuthenticateRequest ),
        HANDLER_MAP_ENTRY( AuthenticateAltRequest ),
        HANDLER_MAP_ENTRY( AuthenticateResponse ),
        HANDLER_MAP_ENTRY( CloseSessionRequest ),
        HANDLER_MAP_ENTRY( CloseSessionResponse ),
        HANDLER_MAP_ENTRY( GetUserIdRequest ),
        HANDLER_MAP_ENTRY( GetUserIdResponse ),
        HANDLER_MAP_ENTRY( GetSessionInfoRequest ),
        HANDLER_MAP_ENTRY( GetSessionInfoResponse ),
    };

#undef HANDLER_MAP_ENTRY

    auto it = funcs.find( type );

    if( it != funcs.end() )
        return it->second( r );

    return nullptr;
}

using basic_parser::MalformedRequest;

request_type_e  detect_request_type( const generic_request::Request & r )
{
    std::string cmd;

    if( r.get_value( "CMD", cmd ) == false )
        throw MalformedRequest( "CMD is not defined" );

    return RequestTypeParser::to_request_type( cmd );
}

} // namespace parser

} // namespace generic_protocol

File: parser.php
<?php

namespace generic_protocol;


// includes
require_once __DIR__.'/../basic_parser/parser.php';

// enums

function parse__ErrorResponse_type_e( & $csv_arr, & $offset )
{
    $res = \basic_parser\parse__int( $csv_arr, $offset );

    return $res;
}

// objects

function parse__SessionInfo( & $csv_arr, & $offset )
{
    $res = new SessionInfo;

    $res->user_id = \basic_parser\parse__int( $csv_arr, $offset );
    $res->start_time = \basic_parser\parse__int( $csv_arr, $offset );
    $res->expiration_time = \basic_parser\parse__int( $csv_arr, $offset );

    return $res;
}

// base messages

function parse__ForwardMessage( & $res, & $csv_arr, & $offset )
{

}

function parse__BackwardMessage( & $res, & $csv_arr, & $offset )
{

}

function parse__Request( & $res, & $csv_arr, & $offset )
{
    $res->session_id = \basic_parser\parse__string( $csv_arr, $offset );

}

// messages

function parse__ErrorResponse( & $csv_arr )
{
    $res = new ErrorResponse;

    $offset = 1;

    // base class
    parse__BackwardMessage( $res, $csv_arr, $offset );

    $res->type = parse__ErrorResponse_type_e( $csv_arr, $offset );
    $res->descr = \basic_parser\parse__string( $csv_arr, $offset );

    return $res;
}

function parse__AuthenticateRequest( & $csv_arr )
{
    $res = new AuthenticateRequest;

    $offset = 1;

    // base class
    parse__ForwardMessage( $res, $csv_arr, $offset );

    $res->user_login = \basic_parser\parse__string( $csv_arr, $offset );
    $res->password = \basic_parser\parse__string( $csv_arr, $offset );

    return $res;
}

function parse__AuthenticateAltRequest( & $csv_arr )
{
    $res = new AuthenticateAltRequest;

    $offset = 1;

    // base class
    parse__ForwardMessage( $res, $csv_arr, $offset );

    $res->user_id = \basic_parser\parse__int( $csv_arr, $offset );
    $res->password = \basic_parser\parse__string( $csv_arr, $offset );

    return $res;
}

function parse__AuthenticateResponse( & $csv_arr )
{
    $res = new AuthenticateResponse;

    $offset = 1;

    // base class
    parse__BackwardMessage( $res, $csv_arr, $offset );

    $res->session_id = \basic_parser\parse__string( $csv_arr, $offset );

    return $res;
}

function parse__CloseSessionRequest( & $csv_arr )
{
    $res = new CloseSessionRequest;

    $offset = 1;

    // base class
    parse__ForwardMessage( $res, $csv_arr, $offset );

    $res->session_id = \basic_parser\parse__string( $csv_arr, $offset );

    return $res;
}

function parse__CloseSessionResponse( & $csv_arr )
{
    $res = new CloseSessionResponse;

    $offset = 1;

    // base class
    parse__BackwardMessage( $res, $csv_arr, $offset );


    return $res;
}

function parse__GetUserIdRequest( & $csv_arr )
{
    $res = new GetUserIdRequest;

    $offset = 1;

    // base class
    parse__Request( $res, $csv_arr, $offset );

    $res->user_login = \basic_parser\parse__string( $csv_arr, $offset );

    return $res;
}

function parse__GetUserIdResponse( & $csv_arr )
{
    $res = new GetUserIdResponse;

    $offset = 1;

    // base class
    parse__BackwardMessage( $res, $csv_arr, $offset );

    $res->user_id = \basic_parser\parse__int( $csv_arr, $offset );

    return $res;
}

function parse__GetSessionInfoRequest( & $csv_arr )
{
    $res = new GetSessionInfoRequest;

    $offset = 1;

    // base class
    parse__Request( $res, $csv_arr, $offset );

    $res->id = \basic_parser\parse__string( $csv_arr, $offset );

    return $res;
}

function parse__GetSessionInfoResponse( & $csv_arr )
{
    $res = new GetSessionInfoResponse;

    $offset = 1;

    // base class
    parse__BackwardMessage( $res, $csv_arr, $offset );

    $res->session_info = parse__SessionInfo( $csv_arr, $offset );

    return $res;
}

// generic

class Parser extends \basic_parser\Parser
{

protected static function parse_csv_array( $csv_arr )
{
    if( sizeof( $csv_arr ) < 1 )
        return self::create_parse_error();

    $handler_map = array(
        // messages
        'generic/ErrorResponse'         => 'parse__ErrorResponse',
        'generic/AuthenticateRequest'         => 'parse__AuthenticateRequest',
        'generic/AuthenticateAltRequest'         => 'parse__AuthenticateAltRequest',
        'generic/AuthenticateResponse'         => 'parse__AuthenticateResponse',
        'generic/CloseSessionRequest'         => 'parse__CloseSessionRequest',
        'generic/CloseSessionResponse'         => 'parse__CloseSessionResponse',
        'generic/GetUserIdRequest'         => 'parse__GetUserIdRequest',
        'generic/GetUserIdResponse'         => 'parse__GetUserIdResponse',
        'generic/GetSessionInfoRequest'         => 'parse__GetSessionInfoRequest',
        'generic/GetSessionInfoResponse'         => 'parse__GetSessionInfoResponse',
    );

    $type = $csv_arr[0][0];

    if( array_key_exists( $type, $handler_map ) )
    {
        $func = '\\generic_protocol\\' . $handler_map[ $type ];
        return $func( $csv_arr[0] );
    }

    return NULL;
}

}

# namespace_end generic_protocol


?>
